clients/gtest/rocblas_gtest_main.cpp

```diff
@@ -209,43 +233,71 @@ static void rocblas_test_sigaction()
     act.sa_flags   = 0;
     act.sa_handler = rocblas_test_signal_handler;
 
-    for(int sig : {SIGABRT, SIGBUS, SIGFPE, SIGILL, SIGPIPE, SIGSEGV, SIGSYS, SIGUSR1, SIGUSR2})
-    {
+    for(int sig : {SIGABRT, SIGBUS, SIGFPE, SIGILL, SIGALRM, SIGSEGV, SIGSYS, SIGUSR1, SIGUSR2})
         sigaction(sig, &act, nullptr);
-    }
 }
 
-// Lambda wrapper which detects signals and exceptions in an invokable function
-void catch_signals_and_exceptions_as_failures(const std::function<void()>& test)
+// Catch std::terminate
+static void terminate_handler()
 {
-    // Save this thread's id, to detect signals in different threads
-    rocblas_test_sighandler_tid = pthread_self();
+    if(state.rocblas_sighandler_enabled)
+        siglongjmp(state.rocblas_sigjmp_buf, -1);
+}
 
-    // Set up the return point
-    int sig = sigsetjmp(rocblas_sigjmp_buf, true);
+// Lambda wrapper which detects signals and exceptions in an invokable function
+void catch_signals_and_exceptions_as_failures(const std::function<void()>& test, bool set_alarm)
+{
+    // Save the current state (to allow nested calls to this function)
+    // Variables must be volatile to survive across sigsetjmp
+    volatile auto old_state     = state;
+    volatile auto old_terminate = std::set_terminate(terminate_handler); // Catch std::terminate
```

Comment:
We allow nested calls to `catch_signals_and_exceptions_as_failures`, catching signals as close as possible to their cause, and thus decreasing the chances of memory leaks.